Math operations

Known4225 opened the pull request:
I've created the Math Operation Benchmarks page and put it under User Guide at the bottom. The page currently only presents the tests and results, it does not elaborate on how to run the tests yourself (I could add if requested, I am still a little unsure about the scope of this project).

npetersen2 commented on source/getting-started/user-guide/math-operations.md:
```diff
@@ -0,0 +1,78 @@
+# Math Operation Benchmarks
+
+Mathematical operations from the <[math.h](https://pubs.opengroup.org/onlinepubs/009696799/basedefs/math.h.html)> library as well as common operations and casts were timed and benchmarked on the AMDC.
```
Looks like an outdated link, please link to:

https://pubs.opengroup.org/onlinepubs/9699919799/

npetersen2 reviewed (changes requested):
Thanks @Known4225 ! Really nice to see the results of your testing, quite interesting.

Your markdown page is a good start on getting this project published. Let's go through a few rounds of review.

Please read and implement the items below. Once you have addressed them, please check them off. Once you are ready for another review from me, please re-tag me in the GitHub interface so I am pinged (and also ping me on Slack).

- [ ] Let's restructure the folders. For pages that have images, the docs website convention is to make a new subfolder for the page, and inside this subfolder, add a `images/` folder and the main page markdown folder. Please do this here. See the `injection/` page for an example of this
- [ ] I see you have many image svg files. Nice. However, the web page only shows 4 files. Can you remove the unneeded files from the repo?
- [ ] Do we need the `consoleOutput.txt` file? I do not see it linked in the doc, so maybe not? If not, please remove.
- [ ] Please try to shorten the page title to be 2 words to match the other `User Guide` pages. This helps with rendering nicely online. Maybe `Math Benchmarks` or `Real-Time Math` or `Math Operations` etc.
- [ ] You seem to use `\` backslash markers at the end of lines of text. Please remove these. If you want stand-alone paragraphs, please insert a new line so it renders as such. I encourage 'nice' typography--take a look at the theme website `furo` for inspiration, https://pradyunsg.me/furo/kitchen-sink/generic/
- [ ] For all math function operations in the text (like `addi` or `sin`), please put them in `code` style so the page is easier to read.
- [ ] Revise the page to be "self-contained." Imagine a new user is trying to figure out the AMDC platform for their research. They need to do complicated math, so they start reading your page. You should write your page to this audience. For example, expand your introduction section, and write a clear takeaways section towards the end. See the GitHub issue (https://github.com/Severson-Group/AMDC-Firmware/issues/395) I wrote for a brief example of some background you can include. Please feel free to copy-paste from the issue into the web page content as you see fit.
- [ ] For your examples of operations you can speed up, please give the C code which implements the speed-up for each. Use the block C environment in markdown so the syntax highlighting works.

Known4225 commented:
- restructured files and folders
- removed unused files (except one that I want to have a toggle button for)
- changed name to be "Math Benchmarks" in files and titles
- removed backslashes
- changed all functions to be `code` style
- added context to introduction. Might want to fix the conclusion, I don't have full larger context
- added code blocks for my faster functions

Additionally, I would like to be able to implement a button that toggles between the sorted and unsorted views of the "All Operations" graph. Not sure if that's outside the realm of possibility with how sphinx works. I haven't looked too much into it yet but I can if you don't know of an existing system that has been used before to do something like this. I could also just add the sorted graph below the unsorted one if this ends up being more trouble than it's worth.

npetersen2 commented:
Thanks @Known4225, it looks like this is ready for my review?

If so, please hit this button so GitHub pings me:

![image](https://github.com/Severson-Group/docs.amdc.dev/assets/20168990/f8651022-2403-4014-96bc-8ca88b4d7b9a)

npetersen2 reviewed (changes requested):
Thanks @Known4225 ! This is looking much better. A few requests below.

I will look into the button to switch between plots and post and update on what I find.

### General requests (mostly on language)

- [x] In the intro section, make these in the `code` environment: "sin(), sqrt(), /"
- [x] The <math.h> links to a page which doesn't seem to describe any math operations? Is this the correct link? Please fix
- [ ] In the `Benchmark Methodology` section: nice! I like how you are giving us the real nuts and bolts of how you profiled each operation. However, this section is still a little hard to read. I recommend the following:
    - [x] Please use the `code` environment liberally. For example, `mathCommandOutputInt` should be in this `code` style. Also, I think all the numbers should be as well.
    - [ ] To help the reader, please add a code block below the bullet list (before `Additional Information:` which gives the C code which implements the above steps. This will help the reader really understand what is going on. This can be a copy-paste from your actual test code, but I encourage you to add comments and make this example very clear, but also succinct.
- [x] There is a statement: "I don't know how this is possible..." yes, ok. It is ok not to understand a result, but please revise the language for two things: 1) remove the "I" language (this is going on a public website where you yourself are not directly attributed as the author, so "I" subject isn't valid); 2) Even though you don't know the answer, please give a potential idea... or, simply say "future analysis is needed to understand this odd behavior" etc.
- [ ] This is an interesting comment: `It was difficult to control caching speedups. Operands to the operations were either constants or declared as register to try to mitigate this problem, but with more complex operations there are many factors that cannot be controlled.` Do you have a concrete example of how changing the code impacts the speed? This could help users write faster code by doing something you suggest, like making things register type or using defines/constants etc
- [x] Please remove "I" language from `...likely factors that I didn’t account for...` and `I am confused by the results of the casting tests`
- [ ] Interesting `fastdivi()` implementation. Are you sure this results in the same exact output for all inputs versus the original?
- [x] For all `Acceleration strategies`, please add a comment above the function (in the code) which gives the original code that is being accelerated. For example, in `fastdivi()`, the comment could say: `// Computes: (arg1 / arg2) for two integers`
- [x] For the end of the article, please add a "Conclusion" section heading before the `Use this page..` part.
- [x] Revise the "feel free to reach out" sentence to instruct the reader to post on the GitHub Discussions, i.e., give them a real actionable item, not just "reach out"
- [x] Please revise the table of contents (i.e., headers) to make more sense. For example, now, there are two "Results" headers, at different levels? And, "Casts" is a top-level header, but isn't this a result? etc.
![image](https://github.com/user-attachments/assets/57a8286d-83f6-48c2-8c36-a575b12c3eee)


### Plot requests

Overall, these are nice plots! Very clean.

- [x] Can you reduce the plot print size so there is less white space between axes labels and tick markers? For example, see below where there is extra space:
![image](https://github.com/user-attachments/assets/809f6862-4504-42c8-9900-e8aeddc74f38)
- [x] Can you annotate this plot with some vertical dashed lines to separate operations with different types? then, add text annotations `ax.annotate()` for each of the 3 sections? See example below:
![image](https://github.com/user-attachments/assets/ea177383-0001-4129-92a6-ea218a9a54be)
- [x] The "All Operations" plot has too much to see on as rendered when I browse the page...see below. I think you can split this into a few plots, e.g., don't show the "common operations" again here to save space, etc.
![image](https://github.com/user-attachments/assets/588ebf98-fbb8-4ed2-8de9-0a1da716191c)
- [ ] The plots are great, but can you add a table of all the results? I am thinking like: 

| Operation | Argument Signature | Time per Operation (ns) | 
| --- | --- | --- |
| `addi` | `int`, `int` | 1.2 | 
| `sqrt` | `double` | 7.5 |
| ... | ... | ... |

npetersen2 commented:
@Known4225 here is a working code snippet for toggling plots from a button on the page:


https://github.com/user-attachments/assets/6fdff481-8c73-4fea-bfa1-34b9783dc1f2

```html
<!-- START SECTION FOR PLOT WITH TOGGLE BUTTON -->

<script>
function toggle_plot() {
    var plot1 = document.getElementById("plot_show_1");
    var plot2 = document.getElementById("plot_show_2");

	plot1.style.display = plot1.style.display === 'none' ? 'block' : 'none';
	plot2.style.display = plot2.style.display === 'none' ? 'block' : 'none';
}
</script>

<button class="sd-badge sd-outline-light sd-text-light" style="cursor:pointer; padding:0.5em;" onclick="toggle_plot()">
Change plot sorting...
</button>

<div id="plot_show_1"  style="display:block;">

![alt text](images/allNanoseconds.svg)

</div>

<div id="plot_show_2"  style="display:none;">

![alt text](images/sortedNanoseconds.svg)

</div>

<!-- END SECTION FOR PLOT WITH TOGGLE BUTTON -->
```

This replaces your `allOperations` SVG plot with a toggle-able version.

----

In short, you can inject raw HTML anywhere in the markdown... as long as true markdown has a empty line between it and the raw html.

Known4225 commented:
I ran some tests to see if declaring operands as `register` would have any impact on performance, but the results came up inconclusive.

After doing 10 tests of 100 operations each, I have calculated these numbers for these four operations:

## Using register operands:
| Operation | Average Ticks per 100 operations | Standard Deviation |
| --- | --- | --- |
| `addi` | 98.3 | 2.452 |
| `subi` | 99.9 | 6.332 |
| `muli` | -90.1 | 7.666 |
| `divi` | 1553.6 | 25.544 |

## No register operands:
| Operation | Average Ticks per 100 operations | Standard Deviation |
| --- | --- | --- |
| `addi` | 97.1 | 4.228 |
| `subi` | 98.8 | 4.962 |
| `muli` | -91.5 | 4.743 |
| `divi` | 1561.1 | 34.598 |

So I can't say whether declaring the operands as `register` had any impact on performance, good or bad. I would definitely not be recommending that users use it in their code, since there are no proven speed benefits and it makes the code more confusing.

codecubepi reviewed (dismissed):
@Known4225 

This looks really good! Definitely some valuable data here. I don't have any requests, but would wait for Nathan's review approval before merging.

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
```diff
@@ -4,21 +4,43 @@ The AMDC is used for real-time control of motor drive systems: every X seconds,
+	baselineTotalTicks += (cpu_timer_now() - tickStart); // record timer
+}
+/* run test on integer add */
+for (register double i = 0; i < iter; i++) {
```
Hmm, is this supposed to be a `double` type for variable `i` ?

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
```diff
@@ -94,6 +232,8 @@ Perhaps the common cause for these operations taking longer is that there is ove
 
+# Conclusion
+
 Use this page as a reference for ballparking about how many of a certain operation will fit within the allocated 100us timeframe. For instance, using the pow() function 140 times would be the around the limit, or doing 4350 integer divisions.
```
Revise this to say:

> The experimental approach and results presented on this page can be used as a reference for ballparking about how many of a certain operation will fit within a time window. For example, given 100 μs of total time, the `pow()` function can be used 140 times, while about 4350 integer divisions can be performed.
>
> Users can evaluate the expected run-time of their code by counting each math operation type and summing the operation time using the tables presented above. Then, users can profile their code on real hardware using similar approaches to above to validate the prediction.

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
````diff
@@ -63,26 +199,28 @@ Quick analysis:
 
 Likewise, why bother using the slow `floor()` function when you can just cast the double to an int and then back to a double.
 ```
+// truncates the decimal portion of a double (note: rounds negative numbers up)
````
Capitalize `Truncates`

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
````diff
@@ -63,26 +199,28 @@ Quick analysis:
 
 One strange result is the slowness of the `hypot()` function, which takes in two lengths and returns the hypotenuse as if the two inputs were sides of a right triangle. This function is abnormally slow, given that the formula is sqrt(x * x + y * y). Which should be around 42.615 nanoseconds but instead it takes on average 181.5 nanoseconds.
 ```
+// calculates the hypotenuse of a right triangle given the lengths of the two legs
````
Capitalize `Calculates`

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
```diff
@@ -35,26 +75,122 @@ Similarily `addf`, `subf`, `mulf`, and `divf` refer to 32-bit floating point ope
+| `castif` | `int` | 3.015 |
+| `castdf` | `double` | 1.545 |
+| `castid` | `int` | 1.575 |
+| `castfd` | `float` | 0.150 |
```
Double tick mark for last row `float`

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/math-benchmarks.md:
````diff
@@ -4,21 +4,43 @@ The AMDC is used for real-time control of motor drive systems: every X seconds,
+- With the measured and baseline numbers of clock cycles logged, the totals are subtracted from each other. The average number of clock cycles is calculated as `(measured - baseline) / 100.0` clock cycles.
+
+Code example:
+```
````
Mark code block with: "```C" and it will help render syntax correctly

npetersen2 commented on source/getting-started/user-guide/math-benchmarks/images/commonNanoseconds.svg:
Nice annotations!

npetersen2 reviewed (changes requested):
Thanks @Known4225 ! This is getting quite close to merging. I went through and left a few comments inline.

Once you have addressed a comment, please click the `Resolve` button.

When you are ready for me to take another look, please re-request my review on the PR.

npetersen2 reviewed (approved):
Looks good! Thanks @Known4225
